Pr mainloop testable

caleridas opened the pull request:
Rework mainloop to make it endpoint behavior decently testable.

After having done this, I am strongly inclined of ripping this sub-par event loop structure out completely and replacing it with something sensible ('asio', 'ace', 'libev', ... maybe 'libtscb' although I feel uncomfortable proposing my own code).

Commit:
Don't re-free timeouts

caleridas committed:
Rework mainloop to allow better testability

Rework mainloop such that it can be created multiple times (in sequence) and
also be torn down (somewhat) correctly. This also requires reworking signal
handlers in a way such that they can be installed/uninstalled properly.

caleridas committed:
Handle endpoint ownership & udp unit test

Make it clear that endpoints are both created and owned by mainloop instance,
rework ownership for them to be more C++-like. Separate freeing of config struct
(which is owned by "main" instead). Expose for testing access purposes.
Verify instantiation in unit test and that they actually send data somewhere.

jkflying commented on src/mavlink-router/mainloop.cpp:
```diff
@@ -37,69 +36,48 @@
-        return -EBUSY;
+    if (instance) {
+        throw std::logic_error("Only one mainloop instance must exist at a given time");
+    }
```
I guess somewhere you need a:

```suggestion
    }
    instance = this;
```

I didn't see it anywhere else

caleridas replied:
yes good catch that is wrong indeed -- I guess it should go to the very end, when all setup is done

jkflying commented on src/mavlink-router/mainloop.cpp:
```diff
@@ -373,6 +299,57 @@ void Mainloop::loop()
+int Mainloop::run_single(int timeout_msec)
+{
+    constexpr int max_events = 8;
+    struct epoll_event events[max_events];
```
I guess this diff section is just a cut-paste-format across from the loop? No other changes?

caleridas replied:
the additional change is pulling out running the mainloop a bit longer for logger shutdown -- the way it was written originally was confusing as hell I think

jkflying commented on src/mavlink-router/mainloop.cpp:
```diff
@@ -381,10 +358,8 @@ bool Mainloop::_log_aggregate_timeout(void *data)
-        for (Endpoint **e = g_endpoints; *e != nullptr; e++) {
-            (*e)->log_aggregate(LOG_AGGREGATE_INTERVAL_SEC);
-        }
+    for (const auto& e : _endpoints) {
```
Way way way cleaner, thank you!

jkflying commented on src/mavlink-router/mainloop.h:
```diff
@@ -59,43 +79,47 @@ class Mainloop {
 
     endpoint_entry *g_tcp_endpoints = nullptr;
-    Endpoint **g_endpoints = nullptr;
+    std::vector<std::unique_ptr<Endpoint>> _endpoints;
```
Much nicer!

jkflying reviewed (approved):
Amazing cleanup. The only thing I saw which doesn't look right is the missing `instance = this;`

jkflying commented:
Ah I didn't notice, the tests are failing :-/

caleridas commented:
> Ah I didn't notice, the tests are failing :-/

In my running it only fails for tests/routing-test.py, but that also failed before this change, I am not clear what the success conditions of this are (it is never run in any of our builds, so maybe it has been broken for long without anyone noticing/caring)

I will have a look what it is and maybe mark it is xfail if needed

caleridas committed:
Make python test skip on missing library

pymavlink.mavutil is neither checked for in configure nor usually installed
in the environments where tests are running. Make the test skip gracefully
if missing instead of causing full test suite failure.

caleridas commented:
Made the python test skip on missing library instead of fail. This does not seem to be running anywhere presently -- arguably it would be nicer to check prerequisites in configure and not declare this a test case, but that is not so easily feasible.
